### PR TITLE
Product collections

### LINKED ISSUE #720: Product collections
Add ability to organize products in collections. Collections should be manageable with the dasboard.

Collection is a set of products that can be in different categories (for example: New arrivals, Winter 2016/2017)

### PR DESCRIPTION
**Pull request by szewczykmira**
Closes #720

### CODE REVIEW

**Comment by artursmet** [comment]
There should be also a view, that will allow to display list of products in the collection on storefront.

File: saleor/dashboard/collection/urls.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from django.conf.urls import url
+
+from . import views
+
+
+urlpatterns = [
+    url(r'^$', views.collection_list, name='collection-list'),
```
Comment (patrys) [inline saleor/dashboard/collection/urls.py]: Please include trailing slashes.

File: saleor/product/migrations/0026_auto_20170102_0927.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            name='Collection',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('name', models.CharField(max_length=128, verbose_name='name', unique=True,)),
+                ('products', models.ManyToManyField(to='product.Product')),
+            ],
+        ),
+        migrations.AlterField(
+            model_name='stocklocation',
+            name='name',
```
Comment (patrys) [inline saleor/product/migrations/0026_auto_20170102_0927.py]: This seems unrelated. Are we missing a migration on `master`?

File: saleor/product/models/base.py [hunk 3]
```diff
@@ -370,3 +370,13 @@ class AttributeChoiceValue(models.Model):
 
     def __str__(self):
         return self.display
+
+
+@python_2_unicode_compatible
+class Collection(models.Model):
+    name = models.CharField(
+        pgettext_lazy('Collection field', 'name'), max_length=128, unique=True)
```
Comment (patrys) [inline saleor/product/models/base.py]: I think `Product collection field` would make it clearer. It's supposed to provide enough context for translators.

**Comment by szewczykmira** [comment]
@patrys django displays warning about this line:
`?: (urls.W002) Your URL pattern '^/$' [name='collection-list'] has a regex beginning with a '/'. Remove this slash as it is unnecessary. If this pattern is targeted in an include(), ensure the include() pattern has a trailing '/'.`
Rest of urls has trailing slashes.

**Comment by patrys** [comment]
Sorry, missed the other ones getting fixed. 🙇

**Comment by patrys** [comment]
Can we do something about test coverage?

**Comment by patrys** [comment]
It looks like migrations need to be regenerated.

**Comment by artursmet** [comment]
@patrys All green and approved, Can we merge it?

**Comment by patrys** [comment]
I'm not sure I follow the requirements here. We already have categories (and soon other taxonomies) for this purpose. I feel we need something entirely different: ability to group similar products so they are listed as alternatives/complementary items on the product details page.

**Comment by artursmet** [comment]
@szewczykmira I updated the ticket according to latest meeting

**Comment by patrys** [comment]
It looks like a rebase gone wrong, seems to contain some unrelated changes.

File: saleor/core/context_processors.py [hunk 4]
```diff
@@ -48,4 +48,4 @@ def webpage_schema(request):
             '@type': 'SearchAction',
             'target': '%s%s?q={search_term}' % (url, reverse('search:search')),
             'query-input': 'required name=search_term'}
-    return {'webpage_schema': json.dumps(data)}
+    return {'webpage_schema': json.dumps(data)}
```
Comment (maciekjablonski) [inline saleor/core/context_processors.py]: new line

Comment (patrys) [inline saleor/core/context_processors.py]: Yes, please set your editor/IDE to always include a trailing newline.

File: saleor/product/models/__init__.py [hunk 5]
```diff
@@ -1,7 +1,8 @@
 from .base import (AttributeChoiceValue, Category, Product, ProductAttribute,
-                   ProductVariant, Stock, ProductClass, StockLocation)
+                   ProductVariant, Stock, ProductClass, StockLocation,
+                   Collection)
 from .images import ProductImage, VariantImage
 
 __all__ = ['AttributeChoiceValue', 'Category', 'Product', 'ProductAttribute',
            'ProductImage', 'ProductVariant', 'Stock', 'VariantImage',
-           'ProductClass', 'StockLocation']
+           'ProductClass', 'StockLocation', 'Collection']
```
Comment (patrys) [inline saleor/product/models/__init__.py]: Can we keep this sorted?

File: saleor/static/js/storefront.js [hunk 6]
```diff
@@ -365,3 +365,6 @@ if ($.cookie('alert') === 'true') {
 $closeMsg.on('click', (e) => {
   $removeProductSucces.addClass('hidden-xs-up');
 });
+
+let $collections = $('#collections');
+$collections.load($collections.data('url'));
```
Comment (patrys) [inline saleor/static/js/storefront.js]: Trailing newline missing.

File: templates/product/_product_collections.html [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  {% for product in products %}
+      <div class="col-6 col-md-4 col-lg-2 product-list" itemscope itemtype="https://schema.org/Product">
+          <a itemprop="url" href="{{ product.get_absolute_url }}">
+              <div class="text-center">
+                  <img itemprop="image" class="img-responsive" src="{% product_first_image product method="crop" size="400x400" %}" alt="">
+              </div>
+          </a>
+      </div>
+  {% endfor %}
+</div>
```
Comment (patrys) [inline templates/product/_product_collections.html]: Newline missing.

**Comment by szewczykmira** [comment]
@artursmet @patrys can you review?

File: templates/product/_product_collections.html [hunk 8]
```diff
@@ -0,0 +1,18 @@
+{% load i18n %}
+{% load staticfiles %}
+{% load price_range from price_ranges %}
+{% load product_first_image from product_images %}
+
+<h2>See also</h2>
```
Comment (krzysztofwolski) [inline templates/product/_product_collections.html]: trans

File: templates/product/_product_collections.html [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+<hr>
+<div class="row">
+  {% for product in products %}
+      <div class="col-6 col-md-4 col-lg-2 product-list" itemscope itemtype="https://schema.org/Product">
```
Comment (krzysztofwolski) [inline templates/product/_product_collections.html]: Update schema to json-ld

File: templates/product/_product_collections.html (hunk 8 above)
Comment (patrys) [inline templates/product/_product_collections.html]: This should likely be the collection name instead.

File: templates/product/_product_collections.html (hunk 9 above)
Comment (patrys) [inline templates/product/_product_collections.html]: Not sure how useful that is since we're going to load it asynchronously. @mociepka?

File: templates/product/_product_collections.html (hunk 8 above)
Comment (szewczykmira) [inline templates/product/_product_collections.html]: It can't be collection name - product can be in many collections so we might display products from different collections.

File: saleor/static/js/storefront.js (hunk 6 above)
Comment (szewczykmira) [inline saleor/static/js/storefront.js]: Done.

File: templates/product/_product_collections.html (hunk 8 above)
Comment (szewczykmira) [inline templates/product/_product_collections.html]: Done

**Comment by szewczykmira** [comment]
@krzysztofwolski @patrys - can you review changes?

File: tests/test_collection.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    assert response.status_code == 200
+
+    collection.refresh_from_db()
+    assert not current_name == collection.name
+    assert list(collection.products.all()) == [product_in_stock]
+
+
+def test_collection_delete_view(admin_client, collection):
+    # Test Http404 when collection doesn't exist
+    pk = collection.pk + 1
```
Comment (krzysztofwolski) [inline tests/test_collection.py]: That's strange. Why not use some big, not existing in db number?

File: saleor/product/urls.py [hunk 11]
```diff
@@ -10,4 +10,6 @@
         views.category_index, name='category'),
     url(r'(?P<slug>[a-z0-9-]+?)-(?P<product_id>[0-9]+)/add/$',
         views.product_add_to_cart, name="add-to-cart"),
+    url(r'^(?P<product_id>[0-9]+)/$',
+        views.products_from_collections, name="from_collections"),
```
Comment (szewczykmira) [inline saleor/product/urls.py]: Done

File: tests/test_collection.py (hunk 10 above)
Comment (szewczykmira) [inline tests/test_collection.py]: Done.

**Comment by elwoodxblues** [comment]
@mad-anne When you finish your current work, we'll have to revisit this PR. It's one year old so I assume there may be some differences, mostly in templates.

File: templates/dashboard/collection/modal_delete.html [hunk 12]
```diff
@@ -0,0 +1,22 @@
+{% load i18n %}
```
Comment (elwoodxblues) [inline templates/dashboard/collection/modal_delete.html]: We have a base modal for such cases: `dashboard/base_modal.html`. Also, we have a convention of naming those templates as `confirm_delete.html`.

File: templates/dashboard/collection/list.html [hunk 13]
```diff
@@ -0,0 +1,70 @@
+{% extends 'dashboard/base.html' %}
```
Comment (elwoodxblues) [inline templates/dashboard/collection/list.html]: Currently, all of our list views in the dashboard provide filtering capabilities, we should add it in this view as well.

File: saleor/product/views.py [hunk 14]
```diff
@@ -1,5 +1,6 @@
 import datetime
 import json
+from random import randint
```
Comment (elwoodxblues) [inline saleor/product/views.py]: Unused import

File: saleor/dashboard/collection/views.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    ctx = {'collections': collections}
+    return TemplateResponse(request,
+                            'dashboard/collection/list.html', ctx)
+
+
+@staff_member_required
+@permission_required('product.edit_product')
+def collection_create(request):
+    collection = Collection()
+    form = CollectionForm(request.POST or None)
```
Comment (elwoodxblues) [inline saleor/dashboard/collection/views.py]: Missing `instance=collection` in the form initialization. I'm not sure now if this is required for a ModelForm to create an object properly, but we follow this rule in other views as well.

File: saleor/dashboard/collection/views.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+from .forms import CollectionForm
+
+
+@staff_member_required
+@permission_required('product.view_product')
+def collection_list(request):
+    collections = Collection.objects.prefetch_related('products').all()
+    collections = get_paginator_items(collections, 30, request.GET.get('page'))
+    ctx = {'collections': collections}
+    return TemplateResponse(request,
```
Comment (elwoodxblues) [inline saleor/dashboard/collection/views.py]: We don't use this line-break style, instead there should be break before `request`. The same applies to the `collection_delete` view.

File: templates/dashboard/collection/list.html (hunk 13 above)
Comment (akjanik) [reply templates/dashboard/collection/list.html]: @elwoodxblues I added option to filter by name, is it enough?

Comment (elwoodxblues) [reply templates/dashboard/collection/list.html]: Yes, for now, we can filter collections only by name.
